zkvm/docs/zkvm-impl.md

```diff
@@ -0,0 +1,75 @@
+
+### Predicates
+
+`Predicate` type represents either an opaque [predicate](zkvm-spec.md#predicate), or an entire tree where unused branches can be left opaque and visited branches are stored as-is. This structure can then produce correct input for the [`call`](zkvm-spec.md#call), [`left`](zkvm-spec.md#left) and [`right`](zkvm-spec.md#right) instructions.
+
+### Variables
+
+Variables are represented as indices into the commitments stored within the VM state. VM manages the "attached" state of the variables to permit replacing the commitments via [`reblind`](zkvm-spec.md#reblind). The secret assignments for the variables are stored within their commitments as [described above](#commitments).
+
+### Expressions
+
+Expressions store their witness data as an optional assignment `Option<ScalarWitness>`.
+The verifier’s VM sees `None` and the prover’s VM sees `Some(...)`.
+
+See also [scalar witness](#scalar-witness).
+
+### Constraints
+
+Constraints do not have explicit witness data. It can be computed on the fly by evaluating the boolean function that the constraint represents, taking the underlying [expressions’](#expressions) assignments as input.
+
+### Signing keys
+
+A [`signtx`](zkvm-spec.md#signtx) instruction expects a [predicate point](zkvm-spec.md#predicate) to be a [verification key](zkvm-spec.md#verification-key). In the `Prover` such key is represented as a `Data::Witness` type that holds `PredicateWitness::Key`. When the prover’s VM pops such item from the stack and remembers it, the `Prover` accumulates all such secret keys and creates a [transaction signature](zkvm-spec.md#transaction-signature) at the end of the execution.
+
+TBD: describe the upcoming signer API that adds a layer of indirection between the witness keys and actual secrets — for HSM-compatibility and multi-party signing scenarios.
+
+### Contracts
+
+An [`input`](#input) instruction decodes a serialized contract. In the prover’s VM it pops an `Input` item from the stack that contains a previously created `Output` object with usual data items (with witnesses) and “frozen values”: values where quantity and flavors are represented by [open commitments](#commitments) instead of variables.
+
+The VM extracts the `Output` object from the `Input` and converts it to a [Contract type](zkvm-spec.md#contract-type) by allocating variables for each commitment within frozen values, turning them into actual [Value](zkvm-spec.md#value-type) types.
+
+### Scalar witness
+
+Scalar witness represents either:
+
+* a [scalar](zkvm-spec.md#scalar), or
+* a [signed integer](../../spacesuit/spec.md#signed-integer)
+
+Arithmetic operations on scalar witnesses _preserve integers until overflow_. If an addition/multiplication of two integers overflows the range of `±(2^64-1)`, the result is promoted to a scalar modulo Ristretto group order.
```

Comment:
does this mean we have to do an overflow check after every arithmetic operation? 
Also this seems like strange behavior... we have to be able to operate over both integers and scalars, but only scalars after integers are overflowed... why not just have the scalar witnesses be represented as scalars all the time instead of doing this conversion? (Or, not allowing for/ throwing an error if an integer overflows? Though that seems like a functionality restriction, it would at least avoid this strange conversion step)

Reply:
The rationale is:

1. Scalars are allowed to overflow because there's no definition of "overflow" for scalars :-)
2. If you integer arithmetic is well in bounds (normally it is), you are good.
3. If you accidentally overflow, rangeproof would fail because you've switched to the Scalar type. So you'll fail to even make a tx with ill-formed smart contract or bad input to it.
4. If you want to explicitly check whether you did overflow early, then you can check the result via `to_integer`, similar to how you check `Option<u64>` after `checked_add`.

Reply:
3. If you overflow, with the rangeproof _proving_ or _verifying_ will fail? If the _proving_ fails upon overflow, then that seems reasonable / safer than the other option.
4. I just question the rationale of allowing overflow (conversion to scalar upon overflow), since it seems like we'd be operating over unnecessarily large numbers at that point... which gives me hesitation for introducing extra complexity to support this edge case

Reply:
>3. If you overflow, will the rangeproof _proving_ or _verifying_ fail? If the _proving_ fails upon overflow, then that seems reasonable / safer than the other option.

Yes, proving.

Reply:
>4. I just question the rationale of allowing overflow (conversion to scalar upon overflow) at all, since it seems like we'd be operating over unnecessarily large numbers at that point... which gives me hesitation for introducing extra complexity to support this edge case

My thinking was that if we just happen to have a "small scalar" (0 or 1) that we want to use with scalars, we can just create ScalarWitness with this number, and it'll become automatically an integer. I felt that forcing it to be a Scalar from the start would make the interface heavier for the user. Also: if we force strict separation between scalars and integers, then we'd need to make Add/Neg/Mul failable (when subtypes cross each other), which may lead to a more ugly usage pattern. Hence, "upgrade int to scalar on overflow" seemed to be a good enough safety solution with none of the usability issues.

(Maybe i'm wrong, though)

Reply:
Talked to @oleganza IRL, summary (for discussion point 4) here:
For operations over integers, we have to have some kind of failure mode for if the integers overflow. Either this can be throwing an error, in which case the functions have to be fallible, which is not ergonomics... or they can convert the result into a scalar, which is the proposal.

I think viewing the `integer -> overflow -> scalar` path as an _error handling_ solution, as opposed to a pathway that is _supposed to_ be used normally, makes a lot more sense to me. And I think that converting to a scalar is a much more user-friendly failure case than making the `add`, `mul` instructions themselves fallible. (though if we find that this is not true in the future, we can revisit this design decision)

Reply:
`add`/`mul` are still usefully "failable" in a sense that one can check what they got like they would check `Option<u64>` from `checked_add`. But their result type is usable in the subsequent operations, unlike Option or Result.